k2/csrc/determinize.cc

```diff
@@ -459,15 +742,229 @@ class DetStateMap {
+    const WfsaWithFbWeights &wfsa_in,
+    float prune_cutoff,
+    vector<Arc> *arcs_out,
+    vector<float> *arc_weights_out,
+    vector<vector<typename TracebackState::DerivType> > *derivs_per_arc,
+    DetStateMap *state_map) {
+  float arc_weight;
+  std::vector<DerivType> deriv_info;
+  state->Normalize(wfsa_in, &arc_weight, &deriv_info);
+  int32_t next_state_id;
+  bool is_new_state = state_map->GetOutputState(state);
+  arcs_out->push_back({this->state_id, next_state_id, label});
+  arc_weights_out->push_back(arc_weight);
+  derivs_per_arc->push_back(std::move(deriv_info));
+  return is_new_state;
+}
+
+
+template <class TracebackState>
+int32_t DetState<TracebackState>::ProcessArcs(
+    const WfsaWithFbWeights &wfsa_in,
+    double prune_cutoff,
+    vector<Arc> *arcs_out,
+    vector<float> *arc_weights_out,
+    vector<vector<typename TracebackState::DerivType> > *derivs_per_arc,
+    DetStateMap *state_map,
+    DetStatePriorityQueue *queue) {
+  int32_t num_steps = 0;
+
+  std::unordered_map<int32_t, DetState<TracebackStats>* > label_to_state;
+
+  // The following loop populates `label_to_state`, creating successor
+  // DetStates (unnormalized).
+  Fsa *fsa = wsfa_in.fsa;
+  const float *arc_weights = wfsa_in.arc_weights;
+  for (const shared_ptr<TracebackState> &state_ptr: elements) {
+    int32_t state_id = state_ptr->state_id,
+        begin_arc = fsa->arc_indexes[state_id],
+        end_arc = fsa->arc_indexes[state_id + 1];
+    num_steps += end_arc - begin_arc;
+    for (int32_t a = begin_arc; a < end_arc; ++a) {
+      const Arc &arc = fsa->arcs[a];
+      float weight = arc_weights[a];
+      int32_t label = arc.label;
+      auto ret = label_to_state.insert({label, nullptr});
+      auto iter = ret.first;
+      if (ret.second) {  // Inserted -> this label was not a key in this map.
+                         // Allocate new DetState.
+        iter->second = new DetState<TracebackState> >(seq_len + 1);
+      }
+      DetState<TracebackState> *det_state = iter->second.get();
+      det_state->Accept(state_ptr, a, arc.label, weight);
+    }
   }
-  float cutoff = a.backward_state_weights[0] - beam;
-  // TODO(dpovey)
+  CHECK(!label_to_state.empty() ||
+        elements[0]->state_id == fsa->FinalState());  // I'm assuming the input
+                                                      // FSA is connected.
+
+  // The following loop normalizes successor det-states, outputs the arcs
+  // that lead to them, and adds them to the queue if necessary.
+  for (auto iter = label_to_state.begin();
+       iter != label_to_state.end(); ++iter) {
+    DetState<TracebackState> *det_state = iter->second;
+
+    float arc_weight;
+    std::vector<DerivType> deriv_info;
+    det_state->Normalize(wfsa_in, &arc_weight, &deriv_info);
+    if (det_state->forward_backward_prob >= prune_cutoff) {
+      bool is_new_state = state_map->GetOutputState(state);
+      arcs_out->push_back({this->state_id, next_state_id, label});
```

Comment:
`GetOutputState(det_state)`, right?
what's the `new_state_id` and `label`

Reply:
yes, GetOutputState(det_state).  next_state_id == state->state_id.  You can get `label` as `iter->first`.